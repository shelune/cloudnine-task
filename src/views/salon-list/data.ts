import { ComponentProps } from "react";
import { SalonListView } from "./salon-list";
import type { SalonItem } from "./salon-row/types";

import salons from "../../assets/data/MOCK_DATA.json";

const MOCK_SALONS: SalonItem[] = salons.slice(0, 100);

// use an external data provider so we can swap out this mock data
// with an actual data source later easier, also better for testing purpose
export const useData = (): ComponentProps<typeof SalonListView> => ({
  salons: MOCK_SALONS,
});
