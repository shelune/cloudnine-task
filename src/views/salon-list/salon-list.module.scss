@import "../../styles/utils/index";

.container {
  height: 100%;
}

.header {
  height: 64px;
  background-color: $white;
  padding: 20px 16px;
  display: flex;
  align-items: center;
  justify-content: space-between;
  border-bottom: 1px solid $pale-gold;

  .title {
    margin: 0;
  }
}

.settingsIcon {
  cursor: pointer;
}

.filterBar {
  padding: 20px 16px;
  width: 100%;
  border-bottom: 1px solid $pale-gold;
}

.filterBarHeader {
  display: flex;
  width: 100%;
  justify-content: space-between;
  align-items: center;
}

.filterBarText {
  @include text-subtitle-md;
}

.priceInput {
  @include text-subtitle-md;
  width: 60px;
  border: none;
  border-bottom: 1px solid $pale-gold;
  border-radius: 0;
  padding: 0;
  outline: none;
  margin-left: 2px;
}

.filterBarToggle {
  @include icon-sm;
  transform: rotate(-90deg);
  cursor: pointer;
  transition: all 0.2s ease-in-out;

  &.isOpen {
    transform: rotate(90deg);
  }
}

.filterBarBody {
  display: flex;
  width: 100%;
  align-items: center;
  transition: all 0.2s ease-in-out;
  max-height: 0;
  overflow: hidden;
  opacity: 0;
  visibility: hidden;

  &.isOpen {
    max-height: 52px;
    opacity: 1;
    visibility: visible;
    padding: 8px 0;
  }
}
.salonListEmpty {
  padding: 16px;
}
