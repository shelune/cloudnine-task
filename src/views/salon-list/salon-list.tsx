import classNames from "classnames";
import React, { FC, useMemo, useState } from "react";
import { Link } from "react-router-dom";

import { ReactComponent as ChevronIcon } from "../../assets/icons/chevron-left.svg";
import { ReactComponent as SettingsIcon } from "../../assets/icons/filters.svg";

import css from "./salon-list.module.scss";
import { SalonRow } from "./salon-row/salon-row";
import type { SalonItem } from "./salon-row/types";

const DEFAULT_MIN_PRICE = 0;
const DEFAULT_MAX_PRICE = 100;

type Props = {
  salons: SalonItem[];
};

export const SalonListView: FC<Props> = ({ salons }) => {
  const [showFilters, setShowFilters] = useState(false);
  const [minPrice, setMinPrice] = useState(DEFAULT_MIN_PRICE);
  const [maxPrice, setMaxPrice] = useState(DEFAULT_MAX_PRICE);

  const filteredSalons = useMemo(
    () =>
      salons.filter(
        (salon) => salon.price >= minPrice && salon.price <= maxPrice,
      ),
    [maxPrice, minPrice, salons],
  );

  return (
    <div className={css.container}>
      <div className={css.header}>
        <div className={css.headerLeft}>
          <Link to="/">
            <ChevronIcon className={css.navIcon} />
          </Link>
        </div>
        <div className={css.headerCenter}>
          <h1 className={css.title}>HÃ¥r</h1>
        </div>
        <div className={css.headerRight}>
          <SettingsIcon className={css.settingsIcon} />
        </div>
      </div>
      <div className={css.filterBar}>
        <div className={css.filterBarHeader}>
          <div className={css.filterBarLeft}>
            <span>{`Price $${minPrice} - $${maxPrice}`}</span>
          </div>
          <div className={css.filterBarRight}>
            <span
              role="button"
              tabIndex={0}
              onKeyDown={() => {}}
              onClick={() => {
                setShowFilters(!showFilters);
              }}
              className={classNames(css.filterBarToggle, {
                [css.isOpen]: showFilters,
              })}
            >
              <ChevronIcon />
            </span>
          </div>
        </div>
        <div
          className={classNames(css.filterBarBody, {
            [css.isOpen]: showFilters,
          })}
        >
          <div className={css.filterBarText}>
            Searching for salons with price between $
          </div>
          <div className={css.minPrice}>
            <input
              className={css.priceInput}
              type="number"
              id="min-price"
              placeholder={`${minPrice}`}
              value={minPrice}
              onChange={(e) => setMinPrice(parseInt(e.target.value, 10))}
              data-testid="min-price"
            />
          </div>
          <div className={css.filterBarText}>and $ </div>
          <div className={css.maxPrice}>
            <input
              className={css.priceInput}
              type="number"
              id="max-price"
              placeholder={`${maxPrice}`}
              value={maxPrice}
              onChange={(e) => setMaxPrice(parseInt(e.target.value, 10))}
              data-testid="max-price"
            />
          </div>
        </div>
      </div>
      <div className={css.salonList}>
        {filteredSalons.length ? (
          filteredSalons.map((salon) => (
            <SalonRow key={salon.id} salon={salon} data-testid="salon-row" />
          ))
        ) : (
          <div className={css.salonListEmpty}>
            Sorry no matching salon found.
          </div>
        )}
      </div>
    </div>
  );
};
