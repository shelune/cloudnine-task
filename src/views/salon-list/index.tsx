import React, { FC, memo } from "react";
import { useData } from "./data";
import { SalonListView as MainComponent } from "./salon-list";

export const SalonListView: FC = memo(
  () => {
    const props = useData();
    // eslint-disable-next-line react/jsx-props-no-spreading
    return <MainComponent {...props} />;
  },
  () => true,
);
