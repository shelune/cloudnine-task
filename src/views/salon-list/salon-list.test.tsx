import React, { ReactNode } from "react";
import { act, fireEvent, render } from "@testing-library/react";
import { createMemoryHistory } from "history";
import { Router } from "react-router-dom";

import { SalonListView } from "./salon-list";
import salons from "../../assets/data/MOCK_DATA.json";

const mockSalons = salons.slice(0, 10);

function renderWithRouter(children: ReactNode) {
  const history = createMemoryHistory();
  return render(
    <Router location={history.location} navigator={history}>
      {children}
    </Router>,
  );
}

describe("GamesView", () => {
  it("should render", () => {
    const { container } = renderWithRouter(
      <SalonListView salons={mockSalons} />,
    );
    expect(container).toMatchSnapshot();
  });

  it("should filter out salons based on price", () => {
    const { getByTestId, getAllByTestId } = renderWithRouter(
      <SalonListView salons={mockSalons} />,
    );
    const minPriceInput = getByTestId("min-price");
    const maxPriceInput = getByTestId("max-price");

    act(() => {
      fireEvent.change(minPriceInput, {
        target: { value: 60 },
      });
      fireEvent.change(maxPriceInput, {
        target: { value: 80 },
      });
    });
    expect(getAllByTestId("salon-row")).toHaveLength(6);
  });

  it("should render empty state when necessary", () => {
    const { container } = renderWithRouter(<SalonListView salons={[]} />);
    expect(container).toMatchSnapshot();
  });
});
