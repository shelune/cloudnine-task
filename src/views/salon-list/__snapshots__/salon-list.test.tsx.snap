// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GamesView should render 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="header"
    >
      <div
        class="headerLeft"
      >
        <a
          href="/"
        >
          <svg
            class="navIcon"
          >
            chevron-left.svg
          </svg>
        </a>
      </div>
      <div
        class="headerCenter"
      >
        <h1
          class="title"
        >
          HÃ¥r
        </h1>
      </div>
      <div
        class="headerRight"
      >
        <svg
          class="settingsIcon"
        >
          filters.svg
        </svg>
      </div>
    </div>
    <div
      class="filterBar"
    >
      <div
        class="filterBarHeader"
      >
        <div
          class="filterBarLeft"
        >
          <span>
            Price $0 - $100
          </span>
        </div>
        <div
          class="filterBarRight"
        >
          <span
            class="filterBarToggle"
            role="button"
            tabindex="0"
          >
            <svg>
              chevron-left.svg
            </svg>
          </span>
        </div>
      </div>
      <div
        class="filterBarBody"
      >
        <div
          class="filterBarText"
        >
          Searching for salons with price between $
        </div>
        <div
          class="minPrice"
        >
          <input
            class="priceInput"
            data-testid="min-price"
            id="min-price"
            placeholder="0"
            type="number"
            value="0"
          />
        </div>
        <div
          class="filterBarText"
        >
          and $ 
        </div>
        <div
          class="maxPrice"
        >
          <input
            class="priceInput"
            data-testid="max-price"
            id="max-price"
            placeholder="100"
            type="number"
            value="100"
          />
        </div>
      </div>
    </div>
    <div
      class="salonList"
    >
      <a
        class="salonRowWrapper"
        href="/salons/2"
      >
        <div
          class="container"
          data-testid="salon-row"
        >
          <div
            class="rowContent"
          >
            <div
              class="rowLeft"
            >
              <span>
                12:00
              </span>
            </div>
            <div
              class="rowCenter"
            >
              <h2
                class="salonName"
              >
                Schuster Group
              </h2>
              <div
                class="salonRating"
              >
                <div
                  class="salonStars"
                >
                  <span
                    class="star"
                  >
                    <svg>
                      star-filled.svg
                    </svg>
                  </span>
                  <span
                    class="star"
                  >
                    <svg>
                      star-empty.svg
                    </svg>
                  </span>
                  <span
                    class="star"
                  >
                    <svg>
                      star-empty.svg
                    </svg>
                  </span>
                  <span
                    class="star"
                  >
                    <svg>
                      star-empty.svg
                    </svg>
                  </span>
                  <span
                    class="star"
                  >
                    <svg>
                      star-empty.svg
                    </svg>
                  </span>
                </div>
                <div
                  class="salonFeedbackCount"
                >
                  (5)
                </div>
              </div>
              <div
                class="salonAddress"
              >
                060 Ramsey Point
              </div>
            </div>
            <div
              class="rowRight"
            >
              <div
                class="salonPrice"
              >
                $70
              </div>
              <div
                class="salonDuration"
              >
                50 min
              </div>
            </div>
          </div>
          <div
            class="rowNav"
          >
            <svg
              class="rowNavIcon"
            >
              chevron-left.svg
            </svg>
          </div>
        </div>
      </a>
      <a
        class="salonRowWrapper"
        href="/salons/3"
      >
        <div
          class="container"
          data-testid="salon-row"
        >
          <div
            class="rowContent"
          >
            <div
              class="rowLeft"
            >
              <span>
                12:00
              </span>
            </div>
            <div
              class="rowCenter"
            >
              <h2
                class="salonName"
              >
                Gottlieb, McLaughlin and Reinger
              </h2>
              <div
                class="salonRating"
              >
                <div
                  class="salonStars"
                >
                  <span
                    class="star"
                  >
                    <svg>
                      star-filled.svg
                    </svg>
                  </span>
                  <span
                    class="star"
                  >
                    <svg>
                      star-filled.svg
                    </svg>
                  </span>
                  <span
                    class="star"
                  >
                    <svg>
                      star-empty.svg
                    </svg>
                  </span>
                  <span
                    class="star"
                  >
                    <svg>
                      star-empty.svg
                    </svg>
                  </span>
                  <span
                    class="star"
                  >
                    <svg>
                      star-empty.svg
                    </svg>
                  </span>
                </div>
                <div
                  class="salonFeedbackCount"
                >
                  (153)
                </div>
              </div>
              <div
                class="salonAddress"
              >
                171 Vera Junction
              </div>
            </div>
            <div
              class="rowRight"
            >
              <div
                class="salonPrice"
              >
                $80
              </div>
              <div
                class="salonDuration"
              >
                30 min
              </div>
            </div>
          </div>
          <div
            class="rowNav"
          >
            <svg
              class="rowNavIcon"
            >
              chevron-left.svg
            </svg>
          </div>
        </div>
      </a>
      <a
        class="salonRowWrapper"
        href="/salons/4"
      >
        <div
          class="container"
          data-testid="salon-row"
        >
          <div
            class="rowContent"
          >
            <div
              class="rowLeft"
            >
              <span>
                21:30
              </span>
            </div>
            <div
              class="rowCenter"
            >
              <h2
                class="salonName"
              >
                Zemlak-Gerlach
              </h2>
              <div
                class="salonRating"
              >
                <div
                  class="salonStars"
                >
                  <span
                    class="star"
                  >
                    <svg>
                      star-filled.svg
                    </svg>
                  </span>
                  <span
                    class="star"
                  >
                    <svg>
                      star-filled.svg
                    </svg>
                  </span>
                  <span
                    class="star"
                  >
                    <svg>
                      star-filled.svg
                    </svg>
                  </span>
                  <span
                    class="star"
                  >
                    <svg>
                      star-empty.svg
                    </svg>
                  </span>
                  <span
                    class="star"
                  >
                    <svg>
                      star-empty.svg
                    </svg>
                  </span>
                </div>
                <div
                  class="salonFeedbackCount"
                >
                  (58)
                </div>
              </div>
              <div
                class="salonAddress"
              >
                1479 Southridge Alley
              </div>
            </div>
            <div
              class="rowRight"
            >
              <div
                class="salonPrice"
              >
                $80
              </div>
              <div
                class="salonDuration"
              >
                50 min
              </div>
            </div>
          </div>
          <div
            class="rowNav"
          >
            <svg
              class="rowNavIcon"
            >
              chevron-left.svg
            </svg>
          </div>
        </div>
      </a>
      <a
        class="salonRowWrapper"
        href="/salons/6"
      >
        <div
          class="container"
          data-testid="salon-row"
        >
          <div
            class="rowContent"
          >
            <div
              class="rowLeft"
            >
              <span>
                10:00
              </span>
            </div>
            <div
              class="rowCenter"
            >
              <h2
                class="salonName"
              >
                Lockman LLC
              </h2>
              <div
                class="salonRating"
              >
                <div
                  class="salonStars"
                >
                  <span
                    class="star"
                  >
                    <svg>
                      star-filled.svg
                    </svg>
                  </span>
                  <span
                    class="star"
                  >
                    <svg>
                      star-filled.svg
                    </svg>
                  </span>
                  <span
                    class="star"
                  >
                    <svg>
                      star-empty.svg
                    </svg>
                  </span>
                  <span
                    class="star"
                  >
                    <svg>
                      star-empty.svg
                    </svg>
                  </span>
                  <span
                    class="star"
                  >
                    <svg>
                      star-empty.svg
                    </svg>
                  </span>
                </div>
                <div
                  class="salonFeedbackCount"
                >
                  (172)
                </div>
              </div>
              <div
                class="salonAddress"
              >
                31 New Castle Way
              </div>
            </div>
            <div
              class="rowRight"
            >
              <div
                class="salonPrice"
              >
                $20
              </div>
              <div
                class="salonDuration"
              >
                40 min
              </div>
            </div>
          </div>
          <div
            class="rowNav"
          >
            <svg
              class="rowNavIcon"
            >
              chevron-left.svg
            </svg>
          </div>
        </div>
      </a>
      <a
        class="salonRowWrapper"
        href="/salons/7"
      >
        <div
          class="container"
          data-testid="salon-row"
        >
          <div
            class="rowContent"
          >
            <div
              class="rowLeft"
            >
              <span>
                12:00
              </span>
            </div>
            <div
              class="rowCenter"
            >
              <h2
                class="salonName"
              >
                Hettinger, Turcotte and Hane
              </h2>
              <div
                class="salonRating"
              >
                <div
                  class="salonStars"
                >
                  <span
                    class="star"
                  >
                    <svg>
                      star-filled.svg
                    </svg>
                  </span>
                  <span
                    class="star"
                  >
                    <svg>
                      star-filled.svg
                    </svg>
                  </span>
                  <span
                    class="star"
                  >
                    <svg>
                      star-empty.svg
                    </svg>
                  </span>
                  <span
                    class="star"
                  >
                    <svg>
                      star-empty.svg
                    </svg>
                  </span>
                  <span
                    class="star"
                  >
                    <svg>
                      star-empty.svg
                    </svg>
                  </span>
                </div>
                <div
                  class="salonFeedbackCount"
                >
                  (152)
                </div>
              </div>
              <div
                class="salonAddress"
              >
                05513 Sutteridge Way
              </div>
            </div>
            <div
              class="rowRight"
            >
              <div
                class="salonPrice"
              >
                $60
              </div>
              <div
                class="salonDuration"
              >
                40 min
              </div>
            </div>
          </div>
          <div
            class="rowNav"
          >
            <svg
              class="rowNavIcon"
            >
              chevron-left.svg
            </svg>
          </div>
        </div>
      </a>
      <a
        class="salonRowWrapper"
        href="/salons/8"
      >
        <div
          class="container"
          data-testid="salon-row"
        >
          <div
            class="rowContent"
          >
            <div
              class="rowLeft"
            >
              <span>
                15:30
              </span>
            </div>
            <div
              class="rowCenter"
            >
              <h2
                class="salonName"
              >
                Hettinger-Brekke
              </h2>
              <div
                class="salonRating"
              >
                <div
                  class="salonStars"
                >
                  <span
                    class="star"
                  >
                    <svg>
                      star-filled.svg
                    </svg>
                  </span>
                  <span
                    class="star"
                  >
                    <svg>
                      star-empty.svg
                    </svg>
                  </span>
                  <span
                    class="star"
                  >
                    <svg>
                      star-empty.svg
                    </svg>
                  </span>
                  <span
                    class="star"
                  >
                    <svg>
                      star-empty.svg
                    </svg>
                  </span>
                  <span
                    class="star"
                  >
                    <svg>
                      star-empty.svg
                    </svg>
                  </span>
                </div>
                <div
                  class="salonFeedbackCount"
                >
                  (64)
                </div>
              </div>
              <div
                class="salonAddress"
              >
                242 Del Mar Alley
              </div>
            </div>
            <div
              class="rowRight"
            >
              <div
                class="salonPrice"
              >
                $60
              </div>
              <div
                class="salonDuration"
              >
                70 min
              </div>
            </div>
          </div>
          <div
            class="rowNav"
          >
            <svg
              class="rowNavIcon"
            >
              chevron-left.svg
            </svg>
          </div>
        </div>
      </a>
      <a
        class="salonRowWrapper"
        href="/salons/10"
      >
        <div
          class="container"
          data-testid="salon-row"
        >
          <div
            class="rowContent"
          >
            <div
              class="rowLeft"
            >
              <span>
                14:00
              </span>
            </div>
            <div
              class="rowCenter"
            >
              <h2
                class="salonName"
              >
                Kilback, O'Connell and Conn
              </h2>
              <div
                class="salonRating"
              >
                <div
                  class="salonStars"
                >
                  <span
                    class="star"
                  >
                    <svg>
                      star-filled.svg
                    </svg>
                  </span>
                  <span
                    class="star"
                  >
                    <svg>
                      star-filled.svg
                    </svg>
                  </span>
                  <span
                    class="star"
                  >
                    <svg>
                      star-filled.svg
                    </svg>
                  </span>
                  <span
                    class="star"
                  >
                    <svg>
                      star-filled.svg
                    </svg>
                  </span>
                  <span
                    class="star"
                  >
                    <svg>
                      star-empty.svg
                    </svg>
                  </span>
                </div>
                <div
                  class="salonFeedbackCount"
                >
                  (8)
                </div>
              </div>
              <div
                class="salonAddress"
              >
                564 Stoughton Pass
              </div>
            </div>
            <div
              class="rowRight"
            >
              <div
                class="salonPrice"
              >
                $60
              </div>
              <div
                class="salonDuration"
              >
                60 min
              </div>
            </div>
          </div>
          <div
            class="rowNav"
          >
            <svg
              class="rowNavIcon"
            >
              chevron-left.svg
            </svg>
          </div>
        </div>
      </a>
    </div>
  </div>
</div>
`;

exports[`GamesView should render empty state when necessary 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="header"
    >
      <div
        class="headerLeft"
      >
        <a
          href="/"
        >
          <svg
            class="navIcon"
          >
            chevron-left.svg
          </svg>
        </a>
      </div>
      <div
        class="headerCenter"
      >
        <h1
          class="title"
        >
          HÃ¥r
        </h1>
      </div>
      <div
        class="headerRight"
      >
        <svg
          class="settingsIcon"
        >
          filters.svg
        </svg>
      </div>
    </div>
    <div
      class="filterBar"
    >
      <div
        class="filterBarHeader"
      >
        <div
          class="filterBarLeft"
        >
          <span>
            Price $0 - $100
          </span>
        </div>
        <div
          class="filterBarRight"
        >
          <span
            class="filterBarToggle"
            role="button"
            tabindex="0"
          >
            <svg>
              chevron-left.svg
            </svg>
          </span>
        </div>
      </div>
      <div
        class="filterBarBody"
      >
        <div
          class="filterBarText"
        >
          Searching for salons with price between $
        </div>
        <div
          class="minPrice"
        >
          <input
            class="priceInput"
            data-testid="min-price"
            id="min-price"
            placeholder="0"
            type="number"
            value="0"
          />
        </div>
        <div
          class="filterBarText"
        >
          and $ 
        </div>
        <div
          class="maxPrice"
        >
          <input
            class="priceInput"
            data-testid="max-price"
            id="max-price"
            placeholder="100"
            type="number"
            value="100"
          />
        </div>
      </div>
    </div>
    <div
      class="salonList"
    >
      <div
        class="salonListEmpty"
      >
        Sorry no matching salon found.
      </div>
    </div>
  </div>
</div>
`;
